### DISCUSSION TITLE
Query builder

### DISCUSSION THREAD

**Discussion by DarkoKukovec** [infinum/datx#517] [v3]
Goal:
* make it easy to support multiple API specs / no specs
  * json:api, hal, etc.
  * custom rest with well defined conventions
  * custom rest without any conventions
  * bonus: support for realtime communication, e.g. web sockets, signalR, firebase, ...
* things to consider
  * headers (including auth)
  * pagination
  * error handling
  * endpoint definition
  * crud operations
  * body / payload
  * network (promise, rx, ...)
  * request body preparation
  * response body parsing

**Comment by DarkoKukovec** [comment]
Here is the first iteration: https://codesandbox.io/s/peaceful-archimedes-lh1i0?file=/src/index.ts

Question: how to abort a request?

**Comment by isBatak** [comment]
Check this out https://github.com/vercel/swr/releases/tag/1.1.0
It's cool how SWR handles key construction, we can do something with that.

Also, SWR handles [conditional fetching](https://swr.vercel.app/docs/conditional-fetching) so I maybe we should not allow nested queries, instead, we should force flat structure.

Instead of this:
```
const resp = client
  .from(User)
  .id("2")
  .request(
    (client, userData) =>
      client
        .from(Post)
        .match({ author: userData.meta.id })
        .request((client, postData) =>
          client
            .from(Comment)
            .match({
              post: postData.map((data) => data.meta.id).join(","),
              user: userData.meta.id
            })
            .request()
        ),
    (client, userData) =>
      client.from(User).match({ friend: userData.meta.id }).request()
  ).swr;
```

we can do this:
```
const { data: userData } = useSWR((client) =>  client.from(User).id("2").key());
const { data: postData } = useSWR((client) =>  client.from(Post).match({ author: userData.meta.id }).key());
const { data: friendData } = useSWR((client) =>   client.from(User).match({ friend: userData.meta.id }).key());
```
- it's flat
- It's easier to handle errors
- it seems more maintainable and reusable
- works better with suspense and error boundaries

**Reply by DarkoKukovec** [reply]
That's great for swr, but I would still want to have this option in case swr is not used.

**Comment by isBatak** [comment]
> Question: how to abort a request?
Here are some examples with https://github.com/vercel/swr/issues/129
I think we can use middleware feature and bake this in SWR hook https://swr.vercel.app/docs/middleware

Something like 
```
function datxMiddleware (useSWRNext) {
  return (key, fetcher, config) => {
    const client = useClient();
    const { key, fetcher, abort } = client.swr(key);
    
    // Handle the next middleware, or the `useSWR` hook if this is the last one.
    const swr = useSWRNext(key, fetcher, config)

    return [swr, abort];
  }
}
```

**Reply by DarkoKukovec** [reply]
We would probably want something that is less swr-centric
